File: misc/ucf/hspt17/sol/wipeout.cpp
#include <bits/stdc++.h>
using namespace std;
#define oo 1000000000L

// our state is defined as <node, <time/distance, stamina>>
typedef pair<int, pair<int, int>> state;

// a struct for our edges
typedef struct Edge {
    int v, d, t;
} Edge;

// the comparator for our priority queue
struct Order {
    bool operator()(state const& a, state const& b) const {
        if(a.second.first == b.second.first)
            return a.second.second >= b.second.second;
        return a.second.first >= b.second.first;
    }
};

// here, we do a simple split-node dijstra's solution
// we split each node (platform) into different ones for each stamina level
// then we take the min of all the different ending staminas for when we reach the end
int main(void) {
    // declare numShows and read in the number of shows
    int numShows;
    scanf("%d", &numShows);
    
    for(int showI = 1; showI <= numShows; showI++) {
        int n, m, s;
        // read in the number of platforms, the number of obstacles, and the starting stamina
        scanf("%d %d %d", &n, &m, &s);
        
        // create our edge list
        vector<Edge> edges[n];
        
        for(int i = 0; i < m; i++) {
            int a, b, d, t;
            // read in the edge and make it zero based
            scanf("%d %d %d %d", &a, &b, &d, &t);
            a--;
            b--;
            
            // create our edge
            Edge e;
            e.v = b;
            e.d = d;
            e.t = t;
            
            // add this edge to our edge list
            edges[a].push_back(e);
        }
        
        // initialize all distances/times to 0
        int dists[n][s + 1];
        for(int i = 0; i < n; i++) 
            for(int j = 0; j <= s; j++)
                dists[i][j] = oo;
        
        // our starting distance for the starting node with full stamina is 0
        dists[0][s] = 0;
        
        // create our priority queue
        priority_queue<state, vector <state>, Order> pq;
        
        // add out initial state
        pq.push(make_pair(0, make_pair(0, s)));
        
        // while there are still states left
        while(!pq.empty()) {
            // pop off the top state
            state st = pq.top();
            pq.pop();
            
            // if we are not the best distance/time, just skip this one
            if(dists[st.first][st.second.second] != st.second.first)
                continue;
            
            // go through each edge from our current platform
            for(Edge e : edges[st.first]) {
                // calculate the new distance/time and the new stamina level
                int newDist = st.second.first + e.t;
                int newStam = st.second.second - e.d;
                
                // if this causes our stamina to go negative or if it doesn't improve out answer,
                // we just skip this one
                if(newStam < 0 || dists[e.v][newStam] <= newDist) continue;
                
                // store our new distance to this next platform
                dists[e.v][newStam] = newDist;
                
                // add this new state to our priority queue
                pq.push(make_pair(e.v, make_pair(newDist, newStam)));
            }
                    
        }
        
        // min all of the distances (times) for all stamina levels
        int ans = oo;
        for(int i = 0; i <= s; i++)
            if(dists[n - 1][i] < ans)
                ans = dists[n - 1][i];
        
        // print our answer
        if(ans == oo) {
            printf("Episode #%d: Wipeout!\n", showI);
        } else {
            printf("Episode #%d: %d\n", showI, ans);
        }
    }
    
    return 0;
}

File: misc/ucf/hspt17/sol/wipeout.py
from heapq import heappush, heappop

INF = 1000000000 # arbitrary big number

T = int(input()) # read number of test cases

for t in range(1, T+1):
	line = input().split(' ') # read whole line
	n = int(line[0]) # set n
	m = int(line[1]) # set m
	s = int(line[2]) # set s

	adj = [] # initialize an adjacency list
	for i in range(n):
		adj.append([]) # make it 2 dimensional
	for i in range(m):
		arr = list(map(int, input().split(' ')))
		# store edge from i to j as a tuple (j, dist, time)
		adj[arr[0]-1].append((arr[1]-1, arr[2], arr[3])) 

	#initialize a 2-D time array representing the shortest amount of time
	# it takes to reach index i with stamina j (represented by time[i][j]
	time = []
	for i in range(n):
		time.append([])
		for j in range(s+1):
			time[i].append(INF)

	#initialize priority queue by starting with S stamina, time = 0, and at index 0
	start = (0,s,0)
	time[0][s] = 0
	pq = []
	heappush(pq,start)
	
	finalTime = INF

	# From here we run standard Dijkstra's algorithm
	while len(pq) != 0:
		node = heappop(pq)
		
		# if at the final platform and time is bettery than our best time, update best
		if(node[2] == n-1 and node[0]<finalTime):
			finalTime = node[0]

		# if we've been to this platform with this stamina before, with less time, just skip it
		if time[node[2]][node[1]] < node[0]:
			continue
		
		# go through all possible obstacles leaving this platform
		for i in adj[node[2]]:
			tS = node[1]-i[1]
			if(tS < 0):
				continue
			newTime = node[0]+i[2]
			
			# if taking this obstacle takes us to the next platform in less time, 
			# take the edge and add it to our priority queue.
			if(time[i[0]][tS] > newTime):
				time[i[0]][tS] = newTime
				heappush(pq,(newTime,tS,i[0]))

	ans = str(finalTime) if finalTime < INF else 'Wipeout!'
	print('Episode #'+str(t)+': '+ans)


